src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py

```diff
@@ -180,20 +259,86 @@ def __init__(
         # store the context in the process of assemble the AgentPool object
         self.context = AKSAgentPoolContext(cmd, raw_parameters, self.models, decorator_mode=DecoratorMode.CREATE)
 
+    def _ensure_agentpool(self, agentpool: AgentPool) -> None:
+        """Internal function to ensure that the incoming `agentpool` object is valid and the same as the attached
+        `agentpool` object in the context.
+
+        If the incomding `agentpool` is not valid or is inconsistent with the `agentpool` in the context, raise a
+        CLIInternalError.
+
+        :return: None
+        """
+        if not isinstance(agentpool, self.models.AgentPool):
+            raise CLIInternalError(
+                "Unexpected agentpool object with type '{}'.".format(type(agentpool))
+            )
+
+        if self.context.agentpool != agentpool:
+            raise CLIInternalError(
+                "Inconsistent state detected. The incoming `agentpool` "
+                "is not the same as the `agentpool` in the context."
+            )
+
     def init_agentpool(self) -> AgentPool:
-        """Initialize an AgentPool object with name and attach it to internal context.
+        """Initialize an AgentPool object and attach it to internal context.
 
         :return: the AgentPool object
         """
-        # Initialize a AgentPool object with name.
-        agentpool = self.models.AgentPool(
-            name=self.context.get_nodepool_name(),
-        )
+        # Initialize a AgentPool object.
+        agentpool = self.models.AgentPool()
```

Comment:
then where's the name is set?

Reply:
no where except in the api path

The `name` is a read only property of the model `AgentPool` and it will be ignored when serialized. I've also checked the code in rp, the name of the nodepool is obtained from the path parameter from api.

Reply:
but you can not predict that one day this value becomes not "read only"?

Reply:
just think whether we can revert it back to fill the name too. because this is not related to your current change.